File: Data-Structure-I/3-Linked-List/P141.cpp
// Date: Sun, Oct 3, 2021 (Not finished)
// Date: Sun, Nov 14, 2021
// 141. Linked List Cycle (https://leetcode.com/problems/linked-list-cycle/)

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */

// Floyd's Hare and Tortoise Algorithm
// Time complexity O(n)
// Space complexity O(1)
bool hasCycle(ListNode *head) {
    if (!head) return head;
    ListNode* slow = head;
    ListNode* fast = head;
    while (fast && fast->next) {
        slow = slow->next;
        fast = fast->next->next;
        if (slow == fast)
            return true;
    }
    return false;
}

File: LeetCoding-Challenge/March-2022/P141.java
package leetcodingchallenge.march2022;

/**
 * Date: March 08, 2022
 * 141. Linked List Cycle
 * https://leetcode.com/problems/linked-list-cycle/
 */

public class P141 {
    // use Floyd's cycle detection algorithm
    // time: O(n)
    // space: O(1)
    public boolean hasCycle(ListNode head) {
        if (head == null) return false;
        ListNode tortoise = head, hore = head.next;
        while (hore != null && hore.next != null) {
            if (tortoise == hore) return true;
            tortoise = tortoise.next;
            hore = hore.next.next;
        }
        return false;
    }

    //Definition for singly-linked list.
    class ListNode {
        int val;
        ListNode next;

        ListNode(int x) {
            val = x;
            next = null;
        }
    }
}
